### PR TITLE
Added checkPackageInstalled and saveDeAsJson

### PR DESCRIPTION
**Pull request by rrydbirk**
As agreed in https://github.com/kharchenkolab/cacoa/issues/34, I added checkPackageInstalled. Further, I added saveDeAsJson so it can be removed from Cacoa and Conos, and I included deview files for this. Finally, I included a bugfix for dotPlot.

### CODE REVIEW

File: R/plots.R [hunk 1]
```diff
@@ -1,14 +1,15 @@
-#' @import ggplot2
+#' @import ggplot2 scales
+#' @importFrom dplyr bind_rows arrange filter select mutate
 #' @importFrom graphics par
 #' @importFrom grDevices adjustcolor rainbow colorRampPalette
 #' @importFrom magrittr %>% %<>% %$%
 #' @importFrom rlang .data
-#' @import scales
```
Comment (VPetukhov) [inline R/plots.R:6]: Are you sure it's never used?

File: R/saveDeAsJson.R [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+#' @examples
+#' \dontrun{
+#' saveDeAsJson(de.raw, sample.groups)
+#' }
+#' ## The results can be viewed in a webbrowser by opening toc.html
+#'
+#' @export
+saveDeAsJson <- function(de.raw, sample.groups=NULL, saveprefix=NULL, dir.name="JSON", gene.metadata=NULL,verbose=TRUE) {
+
+  checkPackageInstalled("jsonlite", cran=TRUE)
```
Comment (VPetukhov) [inline R/saveDeAsJson.R:22]: If you use it here, it should be in Suggests :)

File: R/saveDeAsJson.R [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+         lapply(names(de.raw), function(n)
+           paste0('<tr><td><a href="deview.2.html?d=', saveprefix, make.names(n),'.json">', n, '</a></td></tr>')),
+         list('</table></body></html>')
+  ) %>% paste(collapse='\n')
+
+  write(s, file=toc.file)
+
+  # Copy viewer files
+  file.copy(from = c(paste(system.file(package="sccore"), "extdata/deview.2.html", sep ="/"),
+                     paste(system.file(package="sccore"), "extdata/deview.2.js", sep ="/")), to = paste0(dir.name), overwrite = TRUE)
```
Comment (VPetukhov) [inline R/saveDeAsJson.R]: I'd strongly suspect this not to work on Windows because of "/" in the path... In any case, the [recommended way](https://r-pkgs.org/data.html#filepaths) `system.file("extdata", "deview.2.html", package="sccore")` would be shorter and safer.

**Review by VPetukhov — CHANGES_REQUESTED** [review]
Looks good to me, thank you @rrydbirk! Needs a couple of small fixes for saveDeAsJson.R and that's it.

File: R/saveDeAsJson.R (hunk 3 above)
Comment (VPetukhov) [reply R/saveDeAsJson.R]: Btw, it can also be vectorised as
```r
system.file("extdata", c("deview.2.html", "deview.2.js"), package="sccore")
```
:)

File: R/helpers.R [hunk 4]
```diff
@@ -164,3 +164,40 @@ mergeCountMatrices <- function(cms, transposed=FALSE, ...) {
...
+#' Check for installed package(s)
+#'
+#' @description Check whether a package is installed and suggest how to install from either CRAN or Bioconductor
+#' @param pkgs character Package name(s)
+#' @param details character Helper text (default = "to run this function")
+#' @param install.help character Additional information on how to install package (default = NULL)
+#' @param bioc logical Package(s) is/are available from Bioconductor (default = FALSE)
+#' @param cran logical Package(s) is/are available from CRAN (default = FALSE)
+#' @examples
+#' checkPackageInstalled("sccore", cran = TRUE)
```
Comment (evanbiederstedt) [inline R/helpers.R:177]: This example will probably fail CRAN upon upload....

We could try dontrun as well, but I remember running into issues with this

https://stackoverflow.com/questions/12038160/how-to-not-run-an-example-using-roxygen2

It's probably best just to remove the example (sadly)

File: R/helpers.R [hunk 5]
```diff
@@ -164,3 +164,40 @@ mergeCountMatrices <- function(cms, transposed=FALSE, ...) {
   }
   return(res)
 }
+
+#' Check for installed package(s)
```
Comment (evanbiederstedt) [inline R/helpers.R]: I'd remove this line

Do the following:

```
#'  Check whether a package is installed and suggest how to install from either CRAN or Bioconductor
#'
#' @param pkgs character Package name(s)
```

You'll see that the roxygen2 docs will render correctly

File: R/plots.R [hunk 6]
```diff
@@ -522,7 +523,7 @@ dotPlot <- function (markers,
                      scale.by = "radius",
                      scale.min = NA,
                      scale.max = NA,
-                     verbose=TRUE,
+                     verbose=FALSE,
```
Comment (evanbiederstedt) [inline R/plots.R:526]: Why was this changed?

File: R/saveDeAsJson.R (hunk 3 above)
Comment (evanbiederstedt) [reply R/saveDeAsJson.R]: Could we make these `deview2.html` and `deview.2.js` please?

**Comment by evanbiederstedt** [comment]
@rrydbirk 

(1) Please increase the version in the DESCRIPTION and README as 1.0.3

(2) It's best not to make changes directly in `dev`, but make a branch off of `dev` and do a PR against this. 

I only push to `main` is there's a new version on CRAN; I then release that version

File: R/saveDeAsJson.R (hunk 3 above)
Comment (rrydbirk) [reply R/saveDeAsJson.R]: @evanbiederstedt, what do you mean by "making these", writing them from scratch each time? I've uploaded them in [13400ff](https://github.com/kharchenkolab/sccore/commit/13400ff5a32b054a1dc6096079c03cd83a3d68c5)

File: R/helpers.R (hunk 4 above)
Comment (rrydbirk) [reply R/helpers.R:177]: Could checkPackageInstalled("base") work? Otherwise, let's at least go for dontrun

File: R/plots.R (hunk 6 above)
Comment (rrydbirk) [reply R/plots.R:526]: Honestly, I introduced the progress bar at a time where I was in love with these things. However, in my experience this function is quite fast, no need to show bars per default

**Comment by rrydbirk** [comment]
Thanks @VPetukhov and @evanbiederstedt for review.
About branches, I apologize for my misdemeanor, I'm still not that experienced with this. Should I push the suggested changes to `dev` this time, or...?

File: R/saveDeAsJson.R (hunk 3 above)
Comment (rrydbirk) [reply R/saveDeAsJson.R]: Added Viktor's suggestion

File: R/helpers.R (hunk 5 above)
Comment (rrydbirk) [reply R/helpers.R]: Implemented

File: R/helpers.R (hunk 4 above)
Comment (evanbiederstedt) [reply R/helpers.R:177]: Let's try `dontrun`

If there's a problem, I'll just remove the example

**Review by evanbiederstedt — APPROVED** [review]
LGTM, 
I'll include in the new version. 

I'll close the PR though as I might need to make changes to dev for the release
